File: backend/src/connectors/plugin-types.ts
export type ConnectorAuthType = "oauth2" | "api_key" | "none";

export interface ConnectorRetryPolicy {
  maxAttempts: number;
  backoffMs: number;
  strategy: "fixed" | "exponential";
}

export interface ConnectorManifest {
  key: string;
  displayName: string;
  authType: ConnectorAuthType;
  supportsIncremental: boolean;
  supportsAcl: boolean;
  supportedContentTypes: string[];
  maxPageSize: number;
  retryPolicy: ConnectorRetryPolicy;
  capabilities?: {
    supportsWebhooks?: boolean;
    supportsBinaryContent?: boolean;
    supportsDeleteEvents?: boolean;
    supportsPartialContent?: boolean;
    supportsPerDocumentPermissions?: boolean;
  };
}

export interface ConnectorCheckpoint {
  cursor?: string | null;
  modifiedAfter?: string | null;
  [key: string]: unknown;
}

export interface ConnectorDocumentRef {
  externalId: string;
  title: string | null;
  url: string | null;
  kind: string | null;
  ext: string | null;
  authorName: string | null;
  authorEmail: string | null;
  contentType: string | null;
  sourcePath: string | null;
  sourceLastModifiedAt: string | null;
  sourcePermissions: Record<string, unknown> | null;
  metadata: Record<string, unknown>;
}

export interface ConnectorFetchedDocument {
  content: string | null;
  contentType?: string | null;
  sourcePath?: string | null;
  sourcePermissions?: Record<string, unknown> | null;
  metadata?: Record<string, unknown>;
}

export interface ConnectorPluginContext {
  orgId: string;
  connectorId: string;
  credentials: Record<string, unknown>;
}

export interface ConnectorEnumerateInput extends ConnectorPluginContext {
  config: Record<string, unknown>;
  checkpoint: ConnectorCheckpoint | null;
}

export interface ConnectorFetchInput extends ConnectorPluginContext {
  config: Record<string, unknown>;
  ref: ConnectorDocumentRef;
}

export interface ConnectorEnumerationResult {
  refs: ConnectorDocumentRef[];
  nextCheckpoint: ConnectorCheckpoint | null;
}

export interface ConnectorPlugin {
  validateConfig(config: Record<string, unknown> | null | undefined): Record<string, unknown>;
  enumerate(input: ConnectorEnumerateInput): Promise<ConnectorEnumerationResult>;
  fetchDocument(input: ConnectorFetchInput): Promise<ConnectorFetchedDocument>;
  testConnection?(input: ConnectorPluginContext): Promise<void>;
}

export interface ConnectorAuthProvider {
  authorizeUrl(state: string): string;
  exchangeCode(code: string, redirectUri: string): Promise<Record<string, unknown>>;
  refreshCredentials?(credentials: Record<string, unknown>): Promise<Record<string, unknown> | null>;
}

export interface ConnectorProvider {
  manifest: ConnectorManifest;
  plugin: ConnectorPlugin;
  auth: ConnectorAuthProvider;
}

File: backend/app/connectors/plugin_types.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Literal


ConnectorAuthType = Literal["oauth2", "api_key", "none"]


@dataclass
class ConnectorRetryPolicy:
    max_attempts: int
    backoff_ms: int
    strategy: Literal["fixed", "exponential"]


@dataclass
class ConnectorManifest:
    key: str
    display_name: str
    auth_type: ConnectorAuthType
    supports_incremental: bool
    supports_acl: bool
    supported_content_types: list[str]
    max_page_size: int
    retry_policy: ConnectorRetryPolicy
    capabilities: dict[str, bool] = field(default_factory=dict)


@dataclass
class ConnectorCheckpoint:
    cursor: str | None = None
    modified_after: str | None = None
    extra: dict[str, Any] = field(default_factory=dict)


@dataclass
class ConnectorDocumentRef:
    external_id: str
    title: str | None = None
    url: str | None = None
    kind: str | None = None
    ext: str | None = None
    author_name: str | None = None
    author_email: str | None = None
    content_type: str | None = None
    source_path: str | None = None
    source_last_modified_at: str | None = None
    source_permissions: dict[str, Any] | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class ConnectorFetchedDocument:
    content: str | None
    content_type: str | None = None
    source_path: str | None = None
    source_permissions: dict[str, Any] | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class ConnectorPluginContext:
    org_id: str
    connector_id: str
    credentials: dict[str, Any]


@dataclass
class ConnectorEnumerateInput(ConnectorPluginContext):
    config: dict[str, Any] = field(default_factory=dict)
    checkpoint: ConnectorCheckpoint | None = None


@dataclass
class ConnectorFetchInput(ConnectorPluginContext):
    config: dict[str, Any] = field(default_factory=dict)
    ref: ConnectorDocumentRef = field(default_factory=lambda: ConnectorDocumentRef(""))


@dataclass
class ConnectorEnumerationResult:
    refs: list[ConnectorDocumentRef]
    next_checkpoint: ConnectorCheckpoint | None


class ConnectorPlugin:
    def validate_config(self, config: dict[str, Any] | None) -> dict[str, Any]:
        raise NotImplementedError

    async def enumerate(self, input: ConnectorEnumerateInput) -> ConnectorEnumerationResult:
        raise NotImplementedError

    async def fetch_document(self, input: ConnectorFetchInput) -> ConnectorFetchedDocument:
        raise NotImplementedError


class ConnectorAuthProvider:
    def authorize_url(self, state: str) -> str:
        raise NotImplementedError

    async def exchange_code(self, code: str, redirect_uri: str) -> dict[str, Any]:
        raise NotImplementedError

    async def refresh_credentials(
        self, credentials: dict[str, Any]
    ) -> dict[str, Any] | None:
        return None


@dataclass
class ConnectorProvider:
    manifest: ConnectorManifest
    plugin: ConnectorPlugin
    auth: ConnectorAuthProvider
